### PR TITLE
[10.0][sale_generator] Mig 10 Sale generator with git move

### PR DESCRIPTION
**Pull request by EBII**

### CODE REVIEW

**Review by antoniocanovas — APPROVED** [review]
Tested in #ocacodesprint
Ok functionality.
Good contribution, congratulations !!

File: sale_generator/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# coding: utf-8
+#  @author Sébastien BEAU <sebastien.beau@akretion.com>
+#  @author Abdessamad HILALI <abdessamad.hilali@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (chienandalu) [inline sale_generator/__init__.py]: No need to add credits on init files

File: sale_generator/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# coding: utf-8
+#  @author Sébastien BEAU <sebastien.beau@akretion.com>
+#  @author Abdessamad HILALI <abdessamad.hilali@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (chienandalu) [inline sale_generator/models/__init__.py]: same here

File: sale_generator/tests/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# coding: utf-8
+# 2017 EBII Monsieurb <monsieurb@saaslys.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from . import test_sale_generator
```
Comment (chienandalu) [inline sale_generator/tests/__init__.py]: Same here and add blank line at EOF

**Review by chienandalu — APPROVED** [review]
Functionality is ok. Just some minor stuff proposed

**Review by luismontalba — COMMENTED** [review]
When I try to generate one with this permissions:

imagen

This warning pops up:

imagen

Is it necessary more security configurations?

**Comment by EBII** [comment]
thanks @luismontalba,  i will investigate in security folder ASAP . are you in multicompany mode ?
@chienandalu you right but i see so many i put it in... need to remove ?

**Comment by chienandalu** [comment]
Just on the init files. No need to author them and overall code is cleaner.

File: sale_generator/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    'depends': ['sale_stock'],
+    'data': [
+        'data/data.xml',
+        'views/generator_view.xml',
+        'views/sale_view.xml',
+        'views/partner_view.xml',
+        'security/ir.model.access.csv',
+        'security/ir.rule.csv',
+        ],
+    'demo': [],
```
Comment (Garamotte) [inline sale_generator/__manifest__.py]: Remove empty key in manifest.

File: sale_generator/data/data.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
```
Comment (Garamotte) [inline sale_generator/data/data.xml]: Remove `data` node and move the `noupdate` attribute on the `odoo` node.

File: sale_generator/models/partner.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, api
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    def close_from_customer_wizard(self):
+        id = self.env.context.get('from_generator_id')
```
Comment (Garamotte) [inline sale_generator/models/partner.py]: Please avoid using python's reserved names as variable names.
You can call this `generator_id` for example.

File: sale_generator/models/partner.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        action['res_id'] = id
+        return action
+
+    @api.model
+    def create(self, vals):
+        res = super(ResPartner, self).create(vals)
+        generator_id = self.env.context.get('from_generator_id')
+        if generator_id:
+            generator = self.env['sale.generator'].browse(generator_id)
+            generator.write({'partner_ids': [(4, res.id, 0)]})
```
Comment (Garamotte) [inline sale_generator/models/partner.py]: `generator.partner_ids += res`

File: sale_generator/models/sale_generator.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    sale_ids = fields.One2many(
+        comodel_name='sale.order', inverse_name='generator_id', string="Sales")
+    tmpl_sale_id = fields.Many2one(
+        comodel_name='sale.order',
+        string="Sale Template",
+        required=True,
+        domain=[('is_template', '=', True)])
+    date_order = fields.Datetime(
+        string='Date', oldname='date',
+        default=fields.datetime.now())
```
Comment (Garamotte) [inline sale_generator/models/sale_generator.py]: `fields.Datetime.now()`

File: sale_generator/models/sale_generator.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        ('done', 'Done'),
+    ], string='State', readonly=True, default='draft')
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        string="Company",
+        related="warehouse_id.company_id",
+        store=True)
+
+    @api.multi
+    def _prepare_copy_vals(self, partner):
```
Comment (Garamotte) [inline sale_generator/models/sale_generator.py:42]: Add `self.ensure_one()` when you use directly `self.field_name` statements in the method.

File: sale_generator/models/sale_generator.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        vals = {}
+        vals.update({
```
Comment (Garamotte) [inline sale_generator/models/sale_generator.py]: `vals = {...}`
No need to initialize the variable with an empty dict just to define all values at once on the next line.

File: sale_generator/models/sale_generator.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                    "Can't generate sale order without customer "))
+            else:
+                res.write({'state': 'generating'})
+                res._update_order()
+
+    @api.multi
+    def _update_order(self):
+        self.ensure_one()
+        partner_ids_with_order = [sale.partner_id.id
+                                  for sale in self.sale_ids]
```
Comment (Garamotte) [inline sale_generator/models/sale_generator.py]: I think rewriting this really using the new API would be better, like this (untested):
```python
partners_with_order = self.sale_ids.mapped('partner_id')
for partner in self.partner_ids.filtered(lambda record: record not in partners_with_order):
    self._create_order_for_partner(partner)
for sale in self.sale_ids.filtered(lambda record: record.partner_id not in self.partner_ids):
    sale.unlink()
```

File: sale_generator/models/sale_generator.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+                self._create_order_for_partner(partner)
+        for sale in self.sale_ids:
+            if sale.partner_id not in self.partner_ids:
+                sale.unlink()
+
+    @api.multi
+    def action_confirm(self):
+        for res in self:
+            res.write({'state': 'done'})
+            for sale in res.sale_ids:
```
Comment (Garamotte) [inline sale_generator/models/sale_generator.py]: Is the `for` loop really needed ?
You could use `res.sale_ids.action_confirm()`.

File: sale_generator/sale_generator.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
+from openerp import fields, models, api
```
Comment (Garamotte) [inline sale_generator/sale_generator.py]: I think you forgot to remove this file.

**Review by Garamotte — CHANGES_REQUESTED** [review]

**Comment by EBII** [comment]
@sylvain-garancher thanks for review 
change done

**Comment by EBII** [comment]
@luismontalba it miss a group name for none global rules. see the commit 0217046
https://github.com/OCA/sale-workflow/pull/461/commits/0217046f67543525c6af669fa32142feab6c4215

**Review by luismontalba — APPROVED** [review]

**Comment by bealdav** [comment]
Note for the PSC who'll merge: please squash some useless commits. Thanks

File: sale_generator/tests/test_sale_generator.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
+# coding: utf-8
+# 2017 EBII Monsieurb <monsieurb@saaslys.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.tests.common import TransactionCase
+
+
+class TestProductProfile(TransactionCase):
```
Comment (bealdav) [inline sale_generator/tests/test_sale_generator.py]: It's not ProductProfile here, it seems

File: sale_generator/models/sale_generator.py (hunk 9 above)
Comment (bealdav) [reply sale_generator/models/sale_generator.py:42]: Done

**Review by bealdav — APPROVED** [review]

**Comment by sebastienbeau** [comment]
Merge with a squash here : 2729cb33a426461a769e6feabe620215f3ace1e3

**Pull request closed without merging**